File: example_py/seer_motion_example.py
#!/usr/bin/env python3

import argparse
import logging
import signal
import time

import flexiv_amr


logger = logging.getLogger("seer_motion_example")
stop_requested = False

LINEAR_VELOCITY_X = 0.1
LINEAR_VELOCITY_Y = 0.0
ANGULAR_VELOCITY = 0.1
MOTION_DURATION_MS = 3000.0
WAIT_INTERVAL_SECONDS = 0.05


def signal_handler(signum, frame) -> None:
    global stop_requested
    stop_requested = True
    logger.info("Stop signal received.")


def main() -> int:
    logging.basicConfig(
        level=logging.INFO,
        format="[%(asctime)s.%(msecs)03d] [%(levelname)s] %(message)s",
        datefmt="%Y-%m-%d %H:%M:%S",
    )

    parser = argparse.ArgumentParser(
        description="Drive a SEER AMR with one three-second open-loop velocity command."
    )
    parser.add_argument(
        "amr_ip", help="SEER AMR IP address, for example: 192.168.192.5"
    )
    args = parser.parse_args()

    signal.signal(signal.SIGINT, signal_handler)
    signal.signal(signal.SIGTERM, signal_handler)

    options = flexiv_amr.SeerAmrOptions()
    options.host = args.amr_ip
    options.auto_connect = False
    options.auto_reconnect = True

    amr = flexiv_amr.SeerAmrClient(options)
    motion_started = False
    exit_code = 0

    try:
        logger.info("Connecting to SEER AMR: %s ...", args.amr_ip)
        amr.connect()
        logger.info("Connected: %s", amr.is_connected())

        command = flexiv_amr.SeerOpenLoopMotionCommand()
        command.vx = LINEAR_VELOCITY_X
        command.vy = LINEAR_VELOCITY_Y
        command.w = ANGULAR_VELOCITY
        command.duration = MOTION_DURATION_MS

        logger.warning(
            "Starting open-loop motion: vx=%.3f m/s, vy=%.3f m/s, "
            "w=%.3f rad/s, duration=%.0f ms",
            LINEAR_VELOCITY_X,
            LINEAR_VELOCITY_Y,
            ANGULAR_VELOCITY,
            MOTION_DURATION_MS,
        )

        if not amr.execute_open_loop_motion(command):
            raise RuntimeError("SEER AMR rejected the open-loop motion command")
        motion_started = True

        deadline = time.monotonic() + MOTION_DURATION_MS / 1000.0
        while not stop_requested and time.monotonic() < deadline:
            time.sleep(WAIT_INTERVAL_SECONDS)

        logger.info("Motion duration elapsed or stop requested.")

    except flexiv_amr.AmrException as e:
        logger.error("AMR exception: %s", e)
        exit_code = 1
    except Exception as e:
        logger.exception("Unexpected exception: %s", e)
        exit_code = 1
    finally:
        if motion_started:
            try:
                logger.info("Stopping open-loop motion...")
                if not amr.stop_open_loop_motion():
                    logger.error("SEER AMR rejected the stop command")
                    exit_code = 1
            except Exception as e:
                logger.exception(
                    "Failed to stop open-loop motion during cleanup: %s", e
                )
                exit_code = 1

        try:
            if amr.is_connected():
                amr.disconnect()
        except Exception as e:
            logger.exception("Failed to disconnect during cleanup: %s", e)
            exit_code = 1

    if exit_code == 0:
        logger.info("Motion example completed.")
    return exit_code


if __name__ == "__main__":
    raise SystemExit(main())

File: example/seer_motion_example.cpp
/**
 * @example seer_motion_example.cpp
 * This example drives a SEER AMR with one three-second open-loop velocity command.
 */

#include <chrono>
#include <cstdint>
#include <csignal>
#include <iostream>
#include <stdexcept>
#include <string>
#include <thread>

#include <flexiv/amr/exceptions.h>
#include <flexiv/amr/vendor/seer/seer_amr.h>
#include <spdlog/spdlog.h>

namespace {
volatile std::sig_atomic_t g_stop_requested = 0;

constexpr double kLinearVelocityX = 0.1;
constexpr double kLinearVelocityY = 0.0;
constexpr double kAngularVelocity = 0.1;
constexpr std::int64_t kMotionDurationMs = 3000;
constexpr auto kWaitInterval = std::chrono::milliseconds(50);

void SignalHandler(int) {
    g_stop_requested = 1;
}
}  // namespace

int main(int argc, char* argv[]) {
    if (argc < 2) {
        std::cerr << "Usage:\n"
                  << "  " << argv[0] << " <amr_ip>\n\n"
                  << "Example:\n"
                  << "  " << argv[0] << " 192.168.192.5\n";
        return 1;
    }

    std::signal(SIGINT, SignalHandler);
    std::signal(SIGTERM, SignalHandler);

    spdlog::set_pattern("[%Y-%m-%d %H:%M:%S.%e] [%^%l%$] %v");

    flexiv::amr::seer::SeerAmrOptions options;
    options.host = argv[1];
    options.auto_connect = false;

    flexiv::amr::seer::SeerAmrClient amr(options);
    bool motion_started = false;

    try {
        spdlog::info("Connecting to SEER AMR: {} ...", options.host);
        amr.Connect();
        spdlog::info("Connected: {}", amr.IsConnected());

        flexiv::amr::seer::SeerOpenLoopMotionCommand command;
        command.vx = kLinearVelocityX;
        command.vy = kLinearVelocityY;
        command.w = kAngularVelocity;
        command.duration = kMotionDurationMs;

        spdlog::warn("Starting open-loop motion: vx={} m/s, vy={} m/s, w={} rad/s, duration={} ms",
                     kLinearVelocityX,
                     kLinearVelocityY,
                     kAngularVelocity,
                     kMotionDurationMs);

        if (!amr.ExecuteOpenLoopMotion(command)) {
            throw std::runtime_error("SEER AMR rejected the open-loop motion command");
        }
        motion_started = true;

        const auto deadline =
            std::chrono::steady_clock::now() + std::chrono::milliseconds(kMotionDurationMs);
        while (!g_stop_requested && std::chrono::steady_clock::now() < deadline) {
            std::this_thread::sleep_for(kWaitInterval);
        }

        spdlog::info("Stopping open-loop motion...");
        if (!amr.StopOpenLoopMotion()) {
            throw std::runtime_error("SEER AMR rejected the stop command");
        }
        motion_started = false;

        amr.Disconnect();
        spdlog::info("Motion example completed.");
        return 0;
    } catch (const flexiv::amr::AmrException& e) {
        spdlog::error("AMR exception: {}, code: {}", e.what(), static_cast<int>(e.code()));
    } catch (const std::exception& e) {
        spdlog::error("std exception: {}", e.what());
    }

    if (motion_started) {
        try {
            spdlog::warn("Attempting to stop open-loop motion during cleanup...");
            amr.StopOpenLoopMotion();
        } catch (const std::exception& e) {
            spdlog::error("Failed to stop open-loop motion during cleanup: {}", e.what());
        }
    }

    try {
        amr.Disconnect();
    } catch (const std::exception& e) {
        spdlog::error("Failed to disconnect during cleanup: {}", e.what());
    }

    return 1;
}
